### PR TITLE
LLVM and SPIRV-LLVM-Translator pulldown (WW36)

### PR DESCRIPTION
**Pull request by dbudanov-cmplr**
LLVM: llvm/llvm-project@e1ebe47
SPIRV-LLVM-Translator: KhronosGroup/SPIRV-LLVM-Translator@49b08e8

### CODE REVIEW

**Comment by pvchupin** [comment]
@vmaksimo, @MrSidims , can you check please SPIRV patches? Looks like there've been some issues on merge

**Comment by pvchupin** [comment]
@mdtoguchi, can you help on driver issues please?
```
Failed Tests (5):
  Clang :: Driver/amdgpu-openmp-system-arch.c
  Clang :: Driver/amdgpu-openmp-toolchain.c
  Clang :: Driver/fat_archive_amdgpu.cpp
  Clang :: Driver/fat_archive_nvptx.cpp
  Clang :: Driver/openmp-sycl-interop.c
```

**Comment by MrSidims** [comment]
> @vmaksimo, @MrSidims , can you check please SPIRV patches? Looks like there've been some issues on merge

@pvchupin is there is a way to see where particularly the merge conflict has happen? Thanks

**Comment by pvchupin** [comment]
> > @vmaksimo, @MrSidims , can you check please SPIRV patches? Looks like there've been some issues on merge
> 
> @pvchupin is there is a way to see where particularly the merge conflict has happen? Thanks

error: patch failed: llvm-spirv/test/DebugInfo/DebugInfoChecksumCompileUnit.ll:

@dbudanov-cmplr should have more details.

**Comment by MrSidims** [comment]
>> error: patch failed: llvm-spirv/test/DebugInfo/DebugInfoChecksumCompileUnit.ll:

not sure, what went wrong, but the current changes in the test are looking good

**Comment by dbudanov-cmplr** [comment]
> 
Actually the patch for llvm-spirv/test/DebugInfo/DebugInfoChecksumCompileUnit.ll didn't apply well in my WS. and I did the patch excluding DebugInfoChecksumCompileUnit.ll and then add DebugInfoChecksumCompileUnit.ll manually and then amend it into the same commit. Have no idea why git reported the error on this patch.

**Comment by dbudanov-cmplr** [comment]
> > > @vmaksimo, @MrSidims , can you check please SPIRV patches? Looks like there've been some issues on merge
> > 
> > 
> > @pvchupin is there is a way to see where particularly the merge conflict has happen? Thanks
> 
> error: patch failed: llvm-spirv/test/DebugInfo/DebugInfoChecksumCompileUnit.ll:
> 
> @dbudanov-cmplr should have more details.

Actually the patch for llvm-spirv/test/DebugInfo/DebugInfoChecksumCompileUnit.ll didn't apply well in my WS. and I did the patch excluding DebugInfoChecksumCompileUnit.ll and then add DebugInfoChecksumCompileUnit.ll manually and then amend it into the same commit. Have no idea why git reported the error on this patch.

**Comment by pvchupin** [comment]
@dbudanov-cmplr, @mdtoguchi, to reproduce driver issues
```
Failed Tests (5):                            
  Clang :: Driver/amdgpu-openmp-system-arch.c
  Clang :: Driver/amdgpu-openmp-toolchain.c  
  Clang :: Driver/fat_archive_amdgpu.cpp     
  Clang :: Driver/fat_archive_nvptx.cpp      
  Clang :: Driver/sycl-amdgcn-sqrt.cpp       
```
need to pass additional flag to configure to enable NVPTX and AMDGPU BE:
```
python buildbot/configure.py --enable-all-llvm-targets
```

**Comment by bader** [comment]
> @vmaksimo, @MrSidims , can you check please SPIRV patches? Looks like there've been some issues on merge

See https://github.com/intel/llvm/pull/6536#issuecomment-1243783593.
There were patches missing in previous pulldown, this might cause merge issues, which I experience locally when I try to cherry-pick some patches manually.

**Comment by dbudanov-cmplr** [comment]
> > > @vmaksimo, @MrSidims , can you check please SPIRV patches? Looks like there've been some issues on merge
> > 
> > 
> > See [#6536 (comment)](https://github.com/intel/llvm/pull/6536#issuecomment-1243783593). There were patches missing in previous pulldown, this might cause merge issues, which I experience locally when I try to cherry-pick some patches manually.
> 
> @dbudanov-cmplr, @vmaksimo, @MrSidims can you check that?
Isn't it this https://github.com/dbudanov-cmplr/llvm/commit/479cdcc0d1bbe591ffc8727b44d32237af33a257 ?

**Comment by bader** [comment]
> > > > @vmaksimo, @MrSidims , can you check please SPIRV patches? Looks like there've been some issues on merge
> > > 
> > > 
> > > See [#6536 (comment)](https://github.com/intel/llvm/pull/6536#issuecomment-1243783593). There were patches missing in previous pulldown, this might cause merge issues, which I experience locally when I try to cherry-pick some patches manually.
> > 
> > 
> > @dbudanov-cmplr, @vmaksimo, @MrSidims can you check that?
> > Isn't it this [dbudanov-cmplr@479cdcc](https://github.com/dbudanov-cmplr/llvm/commit/479cdcc0d1bbe591ffc8727b44d32237af33a257) ?

It's, but since .clang-tidy has changed with the pulldown, there must be another commit that reverts this change. Could you find too, please?

**Comment by dbudanov-cmplr** [comment]
> > > > > @vmaksimo, @MrSidims , can you check please SPIRV patches? Looks like there've been some issues on merge
> > > > 
> > > > 
> > > > See [#6536 (comment)](https://github.com/intel/llvm/pull/6536#issuecomment-1243783593). There were patches missing in previous pulldown, this might cause merge issues, which I experience locally when I try to cherry-pick some patches manually.
> > > 
> > > 
> > > @dbudanov-cmplr, @vmaksimo, @MrSidims can you check that?
> > > Isn't it this [dbudanov-cmplr@479cdcc](https://github.com/dbudanov-cmplr/llvm/commit/479cdcc0d1bbe591ffc8727b44d32237af33a257) ?
> 
> It's, but since .clang-tidy has changed with the pulldown, there must be another commit that reverts this change. Could you find too, please?
There are no other commits included in WW33 PR related to clang-tidi https://github.com/dbudanov-cmplr/llvm/commits/479cdcc0d1bbe591ffc8727b44d32237af33a257/.clang-tidy

**Comment by bader** [comment]
I expect to see all these commit in here - https://github.com/intel/llvm/commits/sycl/llvm-spirv/.clang-tidy i.e. in sycl branch.

**Comment by dbudanov-cmplr** [comment]
> I expect to see all these commit in here - https://github.com/intel/llvm/commits/sycl/llvm-spirv/.clang-tidy i.e. in sycl branch.

The latest commit you meant is from Apr 21, 2022, probably it is absent in WW33 pulldown, but it is not likely supposed to be there. Probably I misunderstood the point. Do you mean that 3 commits pointed by  https://github.com/intel/llvm/commits/sycl/llvm-spirv/.clang-tidy are missed in sycl branch and you want them to be integrated?

**Comment by pvchupin** [comment]
HIP and CUDA problem (SYCL :: Regression/fsycl-host-compiler.cpp) fixed in https://github.com/intel/llvm-test-suite/pull/1252

**Comment by pvchupin** [comment]
@bader, thanks for bringing this up. We looked together with @dbudanov-cmplr. It looks like previous pulldown https://github.com/intel/llvm/pull/6536 made a few changes in the wrong places (llvm, rather than llvm-spirv). We made 2 additional changes, on top of fix made by @aelovikov-intel at #6589 to fix this up (https://github.com/intel/llvm/pull/6748/commits/0fe88e826cb42fd6bd2c540eb284f1d413e6216a and https://github.com/intel/llvm/pull/6748/commits/901aa73dcd8014b600011b6acea6b47a428c16d6).
Unfortunately it messes up the history a bit, but currently we should be aligned with the upstream.

**Comment by bader** [comment]
> llvm-spirv/.clang-tidy has a little bit different history due to these issues introduced. We checked that currently it's matching upstream.

I think you mean here is that "**it will match** the upstream after WW36 pulldown will be merged". Right?

**Comment by pvchupin** [comment]
@dbudanov-cmplr, I cherry-picked documentation generation fix
There are these issues left:
* SYCL :: ESIMD/aot_mixed.cpp
* --shared-libs build with clang

**Comment by pvchupin** [comment]
@v-klochkov, @kbobrovs, can you take a look into SYCL :: ESIMD/aot_mixed.cpp one?

**Comment by dbudanov-cmplr** [comment]
> 



> @v-klochkov, @kbobrovs, can you take a look into SYCL :: ESIMD/aot_mixed.cpp one?

@aelovikov-intel could you check if this is the same issues as in https://github.com/intel/llvm-test-suite/pull/1256?

**Comment by aelovikov-intel** [comment]
> > 
> 
> > @v-klochkov, @kbobrovs, can you take a look into SYCL :: ESIMD/aot_mixed.cpp one?
> 
> @aelovikov-intel could you check if this is the same issues as in [intel/llvm-test-suite#1256](https://github.com/intel/llvm-test-suite/pull/1256)?

Where is the link to the fail output?

**Comment by dbudanov-cmplr** [comment]
> 



> > > 
> > 
> > 
> > > @v-klochkov, @kbobrovs, can you take a look into SYCL :: ESIMD/aot_mixed.cpp one?
> > 
> > 
> > @aelovikov-intel could you check if this is the same issues as in [intel/llvm-test-suite#1256](https://github.com/intel/llvm-test-suite/pull/1256)?
> 
> Where is the link to the fail output?

it was https://github.com/intel/llvm/actions/runs/3041366082/jobs/4898814666

**Comment by aelovikov-intel** [comment]
That looks different to what I'm trying to fix in https://github.com/intel/llvm-test-suite/pull/1256

**Comment by dbudanov-cmplr** [comment]
> @v-klochkov, @kbobrovs, can you take a look into SYCL :: ESIMD/aot_mixed.cpp one?
Talked with @v-klochkov, he suggested to create a JIRA and mark the test as XFAIL and continue pulldown.

**Comment by dbudanov-cmplr** [comment]
> 

The created JIRA is CMPLRLLVM-40167

**Comment by dbudanov-cmplr** [comment]
The PR seems ready for a merge. the remaining test ESIMD/aot_mixed.cpp were excluded as UNSUPPORTED by https://github.com/intel/llvm-test-suite/pull/1260 until the issues will be fixed. @bader, @sergey-semenov could you merge this PR.

**Comment by dbudanov-cmplr** [comment]
@dongkyunahn-intel please take a look on the failures: https://github.com/intel/llvm/actions/runs/3049657750/jobs/4916720777
  
they seem to appear after your https://github.com/intel/llvm-test-suite/pull/1259

**Comment by dbudanov-cmplr** [comment]
@pvchupin do you have an idea who can solve the problem with ESIMD Emu llvm test suite?
It seems like CI/configuration issue.

Error: Unable to find an artifact with the name: sycl_linux_default
Run actions/upload-artifact@v1
Error: Path does not exist /localdisk/github/_work/llvm/llvm/build/results_esimd_emu_default.json
Error: Exit code 1 returned from process: file name '/localdisk/github/bin.2.296.2/Runner.PluginHost', arguments 'action "GitHub.Runner.Plugins.Artifact.PublishArtifact, Runner.Plugins"'.

Do you think we can move forward and merge the PR since this is the only issue which remains.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -435,7 +435,7 @@ _pi_context::getFreeSlotInExistingOrNewPool(ze_event_pool_handle_t &Pool,
                                             size_t &Index, bool HostVisible,
                                             bool ProfilingEnabled) {
   // Lock while updating event pool machinery.
-  std::scoped_lock Lock(ZeEventPoolCacheMutex);
+  std::scoped_lock<pi_mutex> Lock(ZeEventPoolCacheMutex);
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:438]: Why was this needed ? The `pi_mutex` is dervide from std:mutex. Tag @againull who authored this

Comment (aelovikov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:438]: It seems this was done in https://github.com/intel/llvm/pull/6748/commits/afbc09286a832ef04cc55020b4c548fe2dfb051d to silence the -Wctad-maybe-unsupported warning, i.e. not a functional requirement.

Comment (pvchupin) [reply sycl/plugins/level_zero/pi_level_zero.cpp:438]: Correct. std::mutex may work here as well if that's better.

**Comment by pvchupin** [comment]
> @pvchupin do you have an idea who can solve the problem with ESIMD Emu llvm test suite? It seems like CI/configuration issue.
> 
> Error: Unable to find an artifact with the name: sycl_linux_default Run actions/upload-artifact@v1 Error: Path does not exist /localdisk/github/_work/llvm/llvm/build/results_esimd_emu_default.json Error: Exit code 1 returned from process: file name '/localdisk/github/bin.2.296.2/Runner.PluginHost', arguments 'action "GitHub.Runner.Plugins.Artifact.PublishArtifact, Runner.Plugins"'.
> 
> Do you think we can move forward and merge the PR since this is the only issue which remains.

@dbudanov-cmplr, yes, looks like network glitch. I've restarted.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:438]: I don't think it's better.

Comment (pvchupin) [reply sycl/plugins/level_zero/pi_level_zero.cpp:438]: The "right" way (according to community) to fix that would be to add explicit deduction guides to headers, but that would be mutex header, which is a system (libstdc++) header. This is sort of workaround until system headers do that
... or community agrees that system headers must be excluded (https://reviews.llvm.org/D133425).
Tagging @AaronBallman

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:438]: I am OK with the proposed change, just need to make sure that future cases like this follow it as well. When is the warning triggered? in which build?

Comment (pvchupin) [reply sycl/plugins/level_zero/pi_level_zero.cpp:438]: Warning is added into LLVM build system inside this pulldown.
As soon as we complete this pulldown, pre-commit CI should fail on build if this is not done (under -Werror).
When building locally you need to add --werror to configure.py.

**Comment by pvchupin** [comment]
New fails on ESIMD Emu target:
```
********************
Failed Tests (5):
  SYCL :: dword_atomic_smoke.cpp
  SYCL :: lsc/lsc_fence.cpp
  SYCL :: lsc/lsc_slm.cpp
  SYCL :: lsc/lsc_surf.cpp
  SYCL :: lsc/lsc_usm.cpp

********************
Unexpectedly Passed Tests (1):
  SYCL :: spec_const/spec_const_double.cpp
```

@dongkyunahn-intel, @v-klochkov can you address please?

**Comment by dongkyunahn-intel** [comment]
> New fails on ESIMD Emu target:
> 
> ```
> ********************
> Failed Tests (5):
>   SYCL :: dword_atomic_smoke.cpp
>   SYCL :: lsc/lsc_fence.cpp
>   SYCL :: lsc/lsc_slm.cpp
>   SYCL :: lsc/lsc_surf.cpp
>   SYCL :: lsc/lsc_usm.cpp
> 
> ********************
> Unexpectedly Passed Tests (1):
>   SYCL :: spec_const/spec_const_double.cpp
> ```
> 
> @dongkyunahn-intel, @v-klochkov can you address please?

PR is ready : https://github.com/intel/llvm-test-suite/pull/1264

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:438]: > pre-commit CI should fail on build if this is not done

Sounds good.

**Comment by pvchupin** [comment]
Remaining fails are unrelated.
* ESIMD Emu issue fails in other PRs (mentioned above). 
* Windows issue is addressed in https://github.com/intel/llvm/pull/6790
* opaque pointers are expected to fail today

**Review by bb-sycl — APPROVED** [review]

**Comment by bb-sycl** [comment]
Thu 15 Sep 2022 02:32:17 PM UTC --- Start to merge the commit into sycl branch. It will take several minutes.

**Comment by bb-sycl** [comment]
Thu 15 Sep 2022 02:36:16 PM UTC --- Merge the branch in this PR to base automatically. Will close the PR later.

**Comment by dongkyunahn-intel** [comment]
> @dongkyunahn-intel, @v-klochkov, this one is not covered by the [intel/llvm-test-suite#1264](https://github.com/intel/llvm-test-suite/pull/1264)
> 
> ```
> Unexpectedly Passed Tests (1):
>   SYCL :: spec_const/spec_const_double.cpp
> ```
> 
> Can you take a look?

The unexpected pass was because the test returns '0' for backends that do not support FP64/double type. While ESIMD_EMULATOR can support FP64 type as native double type, the emulator Plug-in implementation answers request for extension info as if it does not support FP64 type. I fixed the problem in following PR.

https://github.com/intel/llvm/pull/6796

With this PR, the test will fail as expected.

**Comment by yanfeng3721** [comment]
Pre-ci RHEL Build fail(http://icl-jenkins.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FPrecommit_Check_User/detail/Precommit_Check_User/12893/pipeline/89/), impact Linux nightly build. 
The related error log is as below:
```
[2022-09-15T14:53:12.539Z] [868/2957] Building CXX object lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DebugLocStream.cpp.o

[2022-09-15T14:53:12.539Z] FAILED: lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DebugLocStream.cpp.o 

```

**Comment by pvchupin** [comment]
@yanfeng3721, thanks! I also noticed that. Triaging...

**Comment by pvchupin** [comment]
It seems gcc7.5 works fine. Should we migrate?

**Comment by pvchupin** [comment]
@yanfeng3721, community kindly reverted the change at https://github.com/llvm/llvm-project/commit/29c841ce93e087fa4e0c5f3abae94edd460bc24a
It should be pulled in with the next pulldown.
My question on 7.5 migration stays though...
